# Pipeline Guidelines

## Introduction
This pipeline processes retrieved sequence data, curates it into marker-specific FASTA files with taxonomy, optionally concatenates ITS subregions, and performs alignment and taxonomic validation.

## Requirements

### Software
- **R** (â‰¥ 4.0)
- **Python 3** (for `sativa.py`)
- **MAFFT** (sequence alignment tool)

### R Packages
- `data.table`
- `stringr`
- `seqinr`
- `R.utils`

### Commands

```bash
conda create -n plant-curation

conda activate plant-curation

conda install bioconda::itsx
conda install bioconda::mafft

conda install conda-forge::r-base
conda install conda-forge::r-data.table
conda install conda-forge::r-stringr
conda install conda-forge::r-r.utils

conda install bioconda::r-seqinr
```

ACTION - install sativa

### Input
A tab-delimited file (named `download` in the example) with the following fields:
- `nuc`: nucleotide sequence (can include `-` gaps)
- `marker_code`: marker label (in our case: `ITS`, `ITS1`, `ITS2`, `trnL`)
- `kingdom`, `phylum`, `class`, `order`, `family`, `genus`, `species`
- `processid`: unique identifier

Please refer to [`01.data_retrieval.md`](01.data_retrieval.md).

## Pipeline Overview

### Step #1. Data Curation within `R` 
Run the provided R script (`tsv2fasta.R`) to use the raw download file, remove missing/invalid entries, and generate curated FASTA + taxonomy files for ITS and trnL markers.

```bash
Rscript tsv2fasta.R
```

**Output files:**
- `ITS0.fasta.gz`  
- `trnL.fasta.gz`  
- `ITS0.tax`  
- `trnL.tax`  

### Step #2. Running `ITSx`

Run **ITSx** on the input FASTA file to extract and save all ITS subregions using desired number of CPU cores:

```bash
ITSx -i ITS0.fasta -o itsx --cpu {number of threads} --save_regions all
```

**Output files:**
- ACTION - include outputs files from ITSx

### Step #3. ITS Concatenation
Run the provided R script (`construct_ITS_region.R`) to concatenate `ITS1`, `5.8S`, and `ITS2` into a single sequence per sample.

```bash
Rscript construct_ITS_region.R
```

**Output file:**
- `full_ITS.fasta` (concatenated ITS sequences)

ACTION - check if we need to gzip this file

### Step #4. Alignment with `MAFFT`

Align the curated FASTA:

```bash
mafft --auto full_ITS.fasta > full.ITS_aligned.fasta
mafft --auto full_ITS1.fasta > full.ITS1_aligned.fasta
mafft --auto full_ITS2.fasta > full.ITS2_aligned.fasta
mafft --auto trnL_fasta > trnL_aligned.fasta
```

ACTION - check file names

**Output file:**
- `full.ITS_aligned.fasta`
- `trnL_aligned.fasta`

### Step #5. Taxonomic Validation with `SATIVA`

Run [SATIVA](https://github.com/amkozlov/sativa) on the aligned FASTA + taxonomy table:

```bash
python sativa.py -s full_ITS_aligned.fasta -t ITS.tax -T 2 -x bot -n output
python sativa.py -s trnL_aligned.fasta -t trnL.tax -T 2 -x bot -n output
```

**Output files:**
- SATIVA result files (mislabels, reports, etc.)

ACTION - please specify exactly the output files

### Step #6. Filter based on Sativa results

Run the provided R script (`apply_sativa.R`) to correct misclassifications, using sativa .mis output and original .tax file to update it.

ACTION - add commands

## Outputs
Final outputs include:
- Curated FASTA + taxonomy tables (`*.fasta.gz`, `*.tax`)

ACTION - include specific file output names.

## Notes & Tips
- File names are flexible; update script paths accordingly.  
- Ensure MAFFT and SATIVA are available in `$PATH`.  
- Large datasets may require additional memory/CPU.  
- SATIVA `-T` option controls number of threads.  
